lib/src/source/hosted.dart

```diff
@@ -946,80 +923,140 @@ class BoundHostedSource extends CachedSource {
-/// with "--offline".
-///
-/// This uses the system cache to get the list of available packages and does
-/// no network access.
-class _OfflineHostedSource extends BoundHostedSource {
-  _OfflineHostedSource(HostedSource source, SystemCache systemCache)
-      : super(source, systemCache);
+/// The [PackageName.description] for a [HostedSource], storing the package name
+/// and resolved URI of the package server.
+class HostedDescription extends Description<HostedDescription> {
+  final String packageName;
+  final String url;
 
-  /// Gets the list of all versions of [ref] that are in the system cache.
-  @override
-  Future<List<PackageId>> doGetVersions(
-    PackageRef ref,
-    Duration? maxAge,
-  ) async {
-    var parsed = source._asDescription(ref.description);
-    var server = parsed.uri;
-    log.io('Finding versions of ${ref.name} in '
-        '$systemCacheRoot/${_urlToDirectory(server)}');
-
-    var dir = p.join(systemCacheRoot, _urlToDirectory(server));
-
-    List<PackageId> versions;
-    if (dirExists(dir)) {
-      versions = listDir(dir)
-          .where(_looksLikePackageDir)
-          .map((entry) => _idForBasename(p.basename(entry), url: server))
-          .where((id) => id.name == ref.name && id.version != Version.none)
-          .toList();
-    } else {
-      versions = [];
-    }
+  HostedDescription(this.packageName, this.url);
 
-    // If there are no versions in the cache, report a clearer error.
-    if (versions.isEmpty) {
-      throw PackageNotFoundException(
-        'could not find package ${ref.name} in cache',
-        hint: 'Try again without --offline!',
-      );
-    }
+  @override
+  int get hashCode => Object.hash(packageName, url);
 
-    return versions;
+  @override
+  bool operator ==(Object other) {
+    return other is HostedDescription &&
+        other.packageName == packageName &&
+        other.url == url;
   }
 
   @override
-  Future _download(PackageId id, String destPath) {
-    // Since HostedSource is cached, this will only be called for uncached
-    // packages.
-    throw UnsupportedError('Cannot download packages when offline.');
-  }
+  String format() => 'on $url';
 
   @override
-  Future<Pubspec> describeUncached(PackageId id) {
-    throw PackageNotFoundException(
-      '${id.name} ${id.version} is not available in cache',
-      hint: 'Try again without --offline!',
-    );
+  Object? serializeForPubspec({
+    required String? containingDir,
+    required LanguageVersion languageVersion,
+  }) {
+    if (url == source.defaultUrl) {
+      return null;
+    }
+    return {'url': url, 'name': packageName};
   }
 
   @override
-  Future<PackageStatus> status(PackageId id, {Duration? maxAge}) async {
-    // Do we have a cached version response on disk?
-    final versionListing = await _cachedVersionListingResponse(id.toRef());
+  HostedSource get source => HostedSource.instance;
+}
+
+class ResolvedHostedDescription extends ResolvedDescription<HostedDescription> {
```

Comment:
Just curious, but this looks an awful lot like `HostedDescription`, is there any reasons we need `ResolvedDescription` and `Description`?

I'm not suggesting we refactor more now. But if it doesn't make sense to have both of them, could we consider leaving a TODO for future refactoring?

-------------------

Hmm, I see that in `GitResolvedDescription` we have `resolved-ref`, so presumably we need both because this actually the place we could store the `content-hash` is that, right? (just curious).